src/shared/vscode/commands2.ts

```diff
@@ -0,0 +1,330 @@
+            }
+
+            return new CommandResource<Callback>({ id, factory: throwOnRegister }, this.commands)
+        }
+    }
+
+    /**
+     * Registers a new command with the VS Code API.
+     */
+    public register<T extends Callback>(id: string, callback: T): RegisteredCommand<T> {
+        const resource = new CommandResource({ id, factory: () => callback }, this.commands)
+
+        return this.addResource(resource).register()
+    }
+
+    /**
+     * Declares the _intent_ to register a command.
+     *
+     * Forward declaration adds one level of mis-direction. This allows for explicit annotation of
+     * not just the command signature but also its immediate dependencies.
+     */
+    public declare<T extends Callback, D extends any[]>(
+        id: string | { id: string; name: string },
+        factory: CommandFactory<T, D>
+    ): DeclaredCommand<T, D> {
+        const resource = typeof id === 'string' ? { id, factory } : { ...id, factory }
+
+        return this.addResource(new CommandResource(resource, this.commands))
+    }
+
+    /**
+     * Convenience method to declare commands directly from a class.
+     *
+     * Mostly used when many commands are associated with the same long-lived context.
+     * Functions are currently bound at declaration time for simplicity, though this
+     * may change in the future.
+     *
+     * #### Example
+     * ```ts
+     * class Foo {
+     *     public square(n: number): number {
+     *         return n * n
+     *     }
+     * }
+     *
+     * const square = Commands.instance.from(Foo).declareSquare('aws.square')
+     * await square.register(new Foo()).execute(5)
+     * ```
+     */
+    public from<T>(target: new (...args: any[]) => T): Declarables<T> {
+        const result = {} as Record<string, (id: string) => DeclaredCommand>
+
+        for (const [k, v] of Object.entries<Callback>(getFunctions(target))) {
+            const mappedKey = `declare${capitalize(k)}`
+            const name = !isNameMangled() ? `${target.name}.${k}` : ''
+
+            result[mappedKey] = id => this.declare({ id, name }, (instance: T) => v.bind(instance))
+        }
+
+        return result as Declarables<T>
+    }
+
+    public dispose(): void {
+        vscode.Disposable.from(...this.resources.values()).dispose()
+    }
+
+    private addResource<T extends Callback, U extends any[]>(resource: CommandResource<T, U>): CommandResource<T, U> {
+        const registered = this.resources.get(resource.id)
+
+        if (registered?.declared) {
+            throw new Error(`Command "${resource.id}" has already been declared by the Toolkit`)
+        }
+
+        this.resources.set(resource.id, resource)
+        return resource
+    }
+
+    static #instance: Commands
+
+    public static get instance(): Commands {
+        return (this.#instance ??= new Commands())
+    }
+}
+
+/**
+ * Short-hand for {@link Commands.get}.
```

Comment:
do we need these shorthand forms? What does the callsite look like with/without? Fewer names/concepts helps decrease the surface area of the interface (and decreases the variations in the wild, which strengthens learning/re-use/discovery).

Reply:
Not sure. The majority of consumers for this API won't need to care that there's a singleton. I'm favoring a singleton model mostly because module-level state can be painful to test. 

Without the shorthand, callsites would look like this:
`Commands.instance.get`
`Commands.instance.declare`
etc.

Which I mean that's what I already did for `settings.ts`. 
Maybe it'd be nicer to just add this onto the `Commands` class as static methods with the same names.

Reply:
> Maybe it'd be nicer to just add this onto the Commands class as static methods with the same names.

sure. or `Commands.get()` could simply do `Commands.instance.get()`, if we are trying to avoid the `Commands.instance` repetition.